draft/XXXX-multipart-upload/README.md

```diff
@@ -0,0 +1,115 @@
+---
+RFC: XXXX
+Author: Roni Choudhury
+Status: draft
+Created: 2020-08-04
+Last-Modified: 2020-08-04
+---
+
+# RFC XXXX: Multipart Table Upload
+
+This RFC describes how large table data can be safely and efficiently uploaded
+to Multinet, using a variation of the common multipart upload strategy common to
+many web server technologies.
+
+## Motivation
+
+Multipart upload is needed for the following reasons:
+
+1. As a practical matter, our hosting platform Heroku does not allow for
+   requests that take longer than thirty seconds to complete. Some moderately
+   sized CSV files already break this threshold, and so called "big data" tables
+   most certainly will. In order to upload such data files, we will need a way
+   to upload them in chunks.
+2. A long-running one-shot upload does not give the end user much feedback on
+   what is going on. Multipart upload would allow us to provide a more detailed
+   progress meter that can improve the user experience.
+3. If a long-running upload is interrupted for any reason, a multipart upload
+   strategy allows for resuming such uploads without losing the time already
+   spent in uploading one more initial chunks of the file.
+
+## Proposals
+
+### Summary of Approach
+
+The basic idea is as follows: the client initiates a new upload session for a
+given file; this action causes the server to create a new collection in the
+database. Then the client issues one or more "chunk" uploads, each sending a
+range of bytes, up to some size limit, to the server. The server saves each
+chunk as a document in the database collection associated with this file upload.
+Finally, the client issues a "finalize" call, which marks the database
+collection as being complete.
+
+The server can then ask for a file-like object associated with the upload. The
+server can then use this object to perform validation, retrieve table rows, and
+settle them in their final home in the server (as is done currently, directly
+with the file contents).
```

Comment:
Don't we need the whole file for validation? How do you plan to get the whole file out of the db from the multiple parts?

Reply:
The idea is that the `UploadReader` class will give you a file-like interface to the chunks. So you start out at position 0 in the "file", which translates to something like "chunk 0, byte 0". Then a `.read()` call would grab as many bytes as needed, possibly loading more chunks as the virtual file position advances past chunk boundaries.

If that class is available, then you can simply call, e.g., `.readline()` on it to extract lines of a CSV file, or `.read()` followed by `json.loads()` to parse it into JSON, or make use of a streaming JSON reader library, or whatever else we're already doing on the request body currently.

Reply:
While I think on smaller files, you could read in the whole file for validation no problem, ideally we'd want to change the way we validate files to stream the data from any data source, including an `UploadReader`. This is mentioned in https://github.com/multinet-app/multinet-server/issues/408.